### DISCUSSION TITLE
Move Diagram Menu To View's submenu

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#1037] [Q&A] [answered]
Hi,
    I do not want to Diagram menu at the top, I want to move the top diagram menu to the submenu of view menu item, Is any method that can implement this?
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/cb5bf896-6ac1-4780-bded-eebb3fb9fe14)

**Comment by martin-fleck-at** [comment]
@wss29 Hi, the menu is created by the `GLSPDiagramMenuContribution` (see https://github.com/search?q=repo%3Aeclipse-glsp%2Fglsp-theia-integration%20GLSPDiagramMenuContribution&type=code). You can rebind that and simply register the menu action items against `CommonMenus.VIEW` (from `@theia/core/lib/browser`) instead of `GLSPDiagramMenus.DIAGRAM`. For more details on contributing menu items, you can have a look at the Theia documentation: https://theia-ide.org/docs/commands_keybindings/#contributing-menu-items.

I think that should yield the result you want.

**Reply by wss29** [reply]
Hi,
   Do your mean I should re-register every menu action item under GLSPDiagramMenus.DIAGRAM  to CommonMenus.VIEW one by one, not direct move GLSPDiagramMenus.DIAGRAM to be one submenu of CommonMenus.VIEW?

**Reply by martin-fleck-at** [reply]
There are definitely several approaches to this. So feel free to move the diagram menu using a custom menu contribution, of course. I think whatever is easier for you and fits to the MenuModelRegistry API.

**Reply by wss29** [reply]
Hi,
    I used glsp last release version 1.0.0, that version does not have GLSPDiagramMenuContribution , just DiagramMenuContribution
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/55a07f25-78f0-462e-8ce8-d96d26df4645)
I rebind to my new VDMDiagramMenuContribution, but get the error. 
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/1dd8ed0d-7413-432c-a7fd-8b642af8615b)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/29205cd6-2ebd-4099-87d7-c48d3dba9e27)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/348abd9a-3178-4efc-8460-176cbd3d1314)
Could DiagramMenuContribution be rebind in the last release version 1.0.0? or what kind of version can I use to implement my rebind. I do not know how long the new version glsp will release.

**Reply by tortmayr** [reply, marked as answer]
Hi @wss29,
in the 1.0.0 Version the `DiagramMenu` is provided by "sprotty-theia" and unfortunatley it is configured in a way that does not allow easy rebinding. What you could do is use Theia's `FilterContribution` API to filter out the unwanted diagram menu and then simply bind your `VDMDiagramMenuContribution`:
```ts
@injectable()
export class DiagramMenuFilterContribution implements FilterContribution {
    registerContributionFilters(registry: ContributionFilterRegistry): void {
        registry.addFilters([MenuContribution], [contribution => !(contribution instanceof DiagramMenuContribution)]);
    }
}
// in your frontend module
export class VDMTheiaFrontendModule extends GLSPTheiaFrontendModule {
      override configure(context: ContainerContext): void {
        super.configure(context);
        context.bind(FilterContribution).to(DiagramMenuFilterContribution);
        context.bind(MenuContribution).to(VDMDiagramMenuFilterContribution);
    }
}

**Reply by wss29** [reply]
Hi
    thanks for your reply. I'm looking forward to the new release version. could it be released on July 2023? If not I will consider downloading RC version

**Reply by wss29** [reply]
Hi,
    I follow these steps and success, but there is a gap in the menu bar after filtering Diagram menu,
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/1f72fc95-e9ba-4a5d-814f-95b8c22ec487)

**Reply by wss29** [reply]
Hi,
    when I filter GLSPLayoutMenuContribution, the gap disappears. is it because GLSPLayoutMenuContribution has items in Diagram menu?
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/883ba290-add9-42a8-9813-c8f7b16f731b)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/7acd2f2c-c093-43d2-aac7-f7f13afc4554)

**Comment by tortmayr** [comment]
Ah yes, that seems to be the problem. Filtering the `LayoutMenuContribution` definitely solves the problem. Alternatively, if you want to move the layout entries to your new submenu you can also override the `configureDiagramLayoutCommands` method in your frontend module and bind a custom menu contribution:

```ts
export class VDMTheiaFrontendModule extends GLSPTheiaFrontendModule {
 override configureDiagramLayoutCommands(context: ContainerContext): void {
        context.bind(CommandContribution).to(GLSPLayoutCommandContribution).inSingletonScope();
        context.bind(KeybindingContribution).to(GLSPLayoutKeybindingContribution).inSingletonScope();
        context.bind(MenuContribution).to(YourMenuContribution).inSingletonScope();

    }
 }
 ```
